File: apps/mobile/ios/Happy/ViewModels/SessionListViewModel.swift
//
//  SessionListViewModel.swift
//  Happy
//
//  Copyright (c) 2024-2026 Enflame Media. All rights reserved.
//

import Foundation
import Combine

/// ViewModel for the session list screen.
///
/// Manages fetching, filtering, and real-time updates of Claude Code sessions.
/// Uses `ObservableObject` for iOS 16 compatibility.
final class SessionListViewModel: ObservableObject {

    // MARK: - Published Properties

    /// All sessions fetched from the server, sorted by most recent first.
    @Published private(set) var sessions: [Session] = []

    /// Whether a network request is in progress.
    @Published private(set) var isLoading: Bool = false

    /// Whether the initial load has completed (used to distinguish empty state from loading).
    @Published private(set) var hasLoaded: Bool = false

    /// The current error message, if any.
    @Published var errorMessage: String?

    /// Whether to show the error alert.
    @Published var showError: Bool = false

    /// The current filter applied to the session list.
    @Published var filter: SessionFilter = .all

    /// Text entered in the search bar.
    @Published var searchText: String = ""

    /// WebSocket connection status.
    @Published private(set) var connectionStatus: SyncConnectionStatus = .disconnected

    // MARK: - Computed Properties

    /// Sessions filtered by the current filter and search text.
    var filteredSessions: [Session] {
        var result = sessions

        // Apply status filter
        switch filter {
        case .all:
            break
        case .active:
            result = result.filter { $0.status == .active }
        case .completed:
            result = result.filter { $0.status == .completed }
        }

        // Apply search filter
        if !searchText.isEmpty {
            result = result.filter {
                $0.title.localizedCaseInsensitiveContains(searchText)
            }
        }

        return result
    }

    /// Active sessions count for badge display.
    var activeSessionCount: Int {
        sessions.filter { $0.isActive }.count
    }

    /// Whether the session list is empty after filtering.
    var isEmptyState: Bool {
        hasLoaded && filteredSessions.isEmpty
    }

    // MARK: - Dependencies

    private let apiService: any APIServiceProtocol
    private let syncService: SyncService
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    /// Creates a new session list view model.
    ///
    /// - Parameters:
    ///   - apiService: The API service for fetching sessions. Defaults to the shared instance.
    ///   - syncService: The sync service for real-time updates. Defaults to the shared instance.
    init(apiService: any APIServiceProtocol = APIService.shared, syncService: SyncService = .shared) {
        self.apiService = apiService
        self.syncService = syncService
        setupSubscriptions()
    }

    // MARK: - Public Methods

    /// Fetches the session list from the server.
    ///
    /// Sets `isLoading` during the request and updates `sessions` on success.
    /// On failure, sets `errorMessage` and `showError`.
    @MainActor
    func loadSessions() async {
        isLoading = true
        errorMessage = nil

        do {
            let fetchedSessions = try await apiService.fetchSessions()
            sessions = fetchedSessions.sorted { $0.updatedAt > $1.updatedAt }
            hasLoaded = true
        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    /// Refreshes the session list (pull-to-refresh).
    @MainActor
    func refresh() async {
        await loadSessions()
    }

    /// Connects to the WebSocket sync service for real-time updates.
    @MainActor
    func connectSync() async {
        do {
            try await syncService.connect()
        } catch {
            #if DEBUG
            print("[SessionListViewModel] Failed to connect sync: \(error)")
            #endif
        }
    }

    /// Disconnects from the WebSocket sync service.
    func disconnectSync() async {
        await syncService.disconnect()
    }

    /// Dismisses the current error.
    @MainActor
    func dismissError() {
        errorMessage = nil
        showError = false
    }

    // MARK: - Private Methods

    /// Sets up Combine subscriptions for real-time session updates.
    private func setupSubscriptions() {
        // Subscribe to session updates from the sync service
        syncService.sessionUpdates
            .receive(on: DispatchQueue.main)
            .sink { [weak self] updatedSession in
                Task { @MainActor in
                    self?.handleSessionUpdate(updatedSession)
                }
            }
            .store(in: &cancellables)

        // Subscribe to connection status changes
        syncService.connectionStatus
            .receive(on: DispatchQueue.main)
            .sink { [weak self] status in
                Task { @MainActor in
                    self?.connectionStatus = status
                }
            }
            .store(in: &cancellables)
    }

    /// Handles a real-time session update from the sync service.
    ///
    /// If the session already exists in the list, it is updated in place.
    /// Otherwise, the new session is inserted at the top.
    ///
    /// - Parameter updatedSession: The session update received from the server.
    @MainActor
    private func handleSessionUpdate(_ updatedSession: Session) {
        if let index = sessions.firstIndex(where: { $0.id == updatedSession.id }) {
            sessions[index] = updatedSession
        } else {
            sessions.insert(updatedSession, at: 0)
        }

        // Re-sort by most recent
        sessions.sort { $0.updatedAt > $1.updatedAt }
    }
}

// MARK: - Session Filter

/// Filter options for the session list.
enum SessionFilter: String, CaseIterable, Identifiable {
    case all = "All"
    case active = "Active"
    case completed = "Completed"

    var id: String { rawValue }
}

File: apps/mobile/android/app/src/main/java/com/enflame/happy/ui/viewmodel/SessionListViewModel.kt
package com.enflame.happy.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.enflame.happy.domain.model.Session
import com.enflame.happy.domain.model.SessionStatus
import com.enflame.happy.domain.repository.SessionRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Filter options for the session list.
 */
enum class SessionFilter(val label: String) {
    ALL("All"),
    ACTIVE("Active"),
    COMPLETED("Completed"),
    ERROR("Error")
}

/**
 * Connection status for the sync service.
 */
enum class ConnectionStatus {
    CONNECTED,
    CONNECTING,
    DISCONNECTED,
    RECONNECTING
}

/**
 * UI state for the session list screen.
 */
data class SessionListUiState(
    val isLoading: Boolean = false,
    val hasLoaded: Boolean = false,
    val isRefreshing: Boolean = false,
    val errorMessage: String? = null,
    val searchQuery: String = "",
    val filter: SessionFilter = SessionFilter.ALL,
    val connectionStatus: ConnectionStatus = ConnectionStatus.DISCONNECTED
)

/**
 * ViewModel for the Sessions List screen.
 *
 * Manages session loading, filtering, searching, and pull-to-refresh.
 * Observes sessions from the repository as a reactive Flow and combines
 * with UI state for filtered/searched results.
 */
@HiltViewModel
class SessionListViewModel @Inject constructor(
    private val sessionRepository: SessionRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(SessionListUiState())
    val uiState: StateFlow<SessionListUiState> = _uiState.asStateFlow()

    /**
     * Raw sessions from the repository, sorted by most recent activity.
     */
    private val allSessions: StateFlow<List<Session>> = sessionRepository.getSessions()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), emptyList())

    /**
     * Filtered and searched sessions for display.
     * Combines the raw session list with the current filter and search query.
     */
    val filteredSessions: StateFlow<List<Session>> = combine(
        allSessions,
        _uiState
    ) { sessions, state ->
        var result = sessions

        // Apply status filter
        result = when (state.filter) {
            SessionFilter.ALL -> result
            SessionFilter.ACTIVE -> result.filter {
                it.status == SessionStatus.ACTIVE || it.status == SessionStatus.IDLE
            }
            SessionFilter.COMPLETED -> result.filter {
                it.status == SessionStatus.COMPLETED
            }
            SessionFilter.ERROR -> result.filter {
                it.status == SessionStatus.DISCONNECTED
            }
        }

        // Apply search filter
        if (state.searchQuery.isNotBlank()) {
            val query = state.searchQuery.lowercase()
            result = result.filter { session ->
                (session.title?.lowercase()?.contains(query) == true) ||
                    (session.machineName?.lowercase()?.contains(query) == true) ||
                    session.id.lowercase().contains(query)
            }
        }

        // Sort by most recent activity
        result.sortedByDescending { it.updatedAt ?: it.createdAt }
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), emptyList())

    init {
        loadSessions()
    }

    /**
     * Initial load of sessions from the repository.
     */
    fun loadSessions() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, errorMessage = null) }
            try {
                sessionRepository.refreshSessions()
                _uiState.update { it.copy(isLoading = false, hasLoaded = true) }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        hasLoaded = true,
                        errorMessage = e.message ?: "Failed to load sessions"
                    )
                }
            }
        }
    }

    /**
     * Refresh sessions via pull-to-refresh.
     */
    fun refreshSessions() {
        viewModelScope.launch {
            _uiState.update { it.copy(isRefreshing = true, errorMessage = null) }
            try {
                sessionRepository.refreshSessions()
                _uiState.update { it.copy(isRefreshing = false) }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isRefreshing = false,
                        errorMessage = e.message ?: "Failed to refresh sessions"
                    )
                }
            }
        }
    }

    /**
     * Update the search query for filtering sessions.
     */
    fun onSearchQueryChanged(query: String) {
        _uiState.update { it.copy(searchQuery = query) }
    }

    /**
     * Update the status filter.
     */
    fun onFilterChanged(filter: SessionFilter) {
        _uiState.update { it.copy(filter = filter) }
    }

    /**
     * Dismiss the current error message.
     */
    fun dismissError() {
        _uiState.update { it.copy(errorMessage = null) }
    }
}
